### PR TITLE
CollisionSystem

### LINKED ISSUE #14: Collision system.

### PR DESCRIPTION
**Pull request by JacopoG**
Refers to #14.

Several things here. At the end I implemented the functionality without Box2D (it include too much features). The collider system has two main responsability:
  1. Update the CollisionComponent state, adding the entities that are colliding with the entity owner of the component.
  2. Allow others to register for collision events and dispatch these events. When you register to an event, you have to specify the families of the two entities you want to listen for collision (for example, you can register for collision events between bomb and player, using the appropriate families).

I edited the CollisionComponent because I needed the `overlaps` features of the `Rectange` and `Circle`. Furthermore, for me make sense to avoid (as best as we can) embedding specific objects in the components. I adopted a more abstract approach were the relevant objects are instantiated inside the system (BUT the code that does this thing makes me feel bad, take a look at the method `areColliding`, no better ideas at the moment). I guess keeping the components simple help on isolation, but not sure. Share your thoughts!

TODO:
  - [ ] handle rotation of the shape

Give me feedback!

```
// example, how to register for an event
engine.getSystem(CollisionSystem::class.java).registerListener(
	allOf(CollisionComponent::class, MovementComponent::class).get(),
	allOf(CollisionComponent::class).get(),
	Listener<CollisionEvent>(function = {_, _ -> println("new collision")})
)
```

### CODE REVIEW

File: core/src/no/group15/playmagic/ecs/systems/CollisionSystem.kt [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+
+
+	// -- IMPLEMENTATION --
+
+
+
+	private fun isNewCollision(entity1: Entity, entity2: Entity): Boolean {
+		return ! entity1[collision]!!.collidingWith.contains(entity2)
```
Comment (skykanin) [inline core/src/no/group15/playmagic/ecs/systems/CollisionSystem.kt]: When you're returning a 1 liner like this you can use `fun isNewCollision(...) = returnValue` and omit the `return` keyword. It's a bit more concise

File: core/src/no/group15/playmagic/ecs/systems/CollisionSystem.kt [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+
+
+	private fun areColliding(entity1: Entity, entity2: Entity): Boolean {
+		var shape1: ColliderShape = entity1[collision]!!.shape
+		var shape2: ColliderShape = entity2[collision]!!.shape
```
Comment (skykanin) [inline core/src/no/group15/playmagic/ecs/systems/CollisionSystem.kt]: These values can be declared as `val`

**Review by skykanin — CHANGES_REQUESTED** [review]
Overall it looks good, just a couple of small changes that can make the code a bit more concise

**Comment by JacopoG** [comment]
@skykanin thank you, I like this kind of suggestions. I committed the changes you proposed, check! The collision system is critical, I think is good if also @DeimosA and @eldarht take a look to it.

**Comment by skykanin** [comment]
Another thing you should do is run the code formatter (`ctrl + alt + l` in android studio) to properly format the code. Optionally if you run it with `ctrl + shift + alt + l` you also get options for code clean-up (removing unused imports etc.)

File: core/src/no/group15/playmagic/ecs/systems/CollisionSystem.kt [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+
+
+	private fun areColliding(entity1: Entity, entity2: Entity): Boolean {
+		val shape1: ColliderShape = entity1[collision]!!.shape
+		val shape2: ColliderShape = entity2[collision]!!.shape
+
+		return when (shape1) {
+
+			ColliderShape.CIRCLE -> when (shape2) {
```
Comment (DeimosA) [inline core/src/no/group15/playmagic/ecs/systems/CollisionSystem.kt]: Do we need to consider circles? tiles are square and don't rotate other than 90 degrees

**Review by DeimosA — APPROVED** [review]
Looks overly complex, so now my brain hurts 🥇

File: core/src/no/group15/playmagic/ecs/systems/CollisionSystem.kt (hunk 3 above)
Comment (JacopoG) [reply core/src/no/group15/playmagic/ecs/systems/CollisionSystem.kt]: Happy to know that, it is difficult to infer this kind of information (about tiles and rotation). I thought we need it for the explosions, but if we use a "discrete" grid, no, we probably don't need it.

**Comment by JacopoG** [comment]
> Looks overly complex, so now my brain hurts 1st_place_medal

@DeimosA I'm sorry about that. I guess you went deeply in the implementation, which it isn't so tidy. My intent was to highlight the public interface (how do you can register to a collision event), to have a opinion on it (does it make sense or not?). If you have better idea or if you can share more detailed information about what look overly complex I'll be happy to get your feedback, seriously. I'm here to learn and nothing else. What can I improve? :)

**Comment by DeimosA** [comment]
Don't be sorry, I was just tired when I looked at it.
I'm just uncertain as to the purpose of generating all the pairs and things
I'd imagine something simple:
```
for ( i in 0 until entities.size()) {
  for ( j in i+1 until entities.size()) {
    if ( /* rectangles overlap */ ) {
      /* send signal */
    }
  }
}
```
Too simple?

**Comment by JacopoG** [comment]
@DeimosA thank you for taking the time to go deeper, I really appreciate that.

The `for` can be done as you suggest, the `entityPairs()` function is born after some frustration with the language (there are problems with nested iterators). But yes, it isn't needed.

I think the body of the `for` should be more complex. If you don't handle the state, you don't know which are new collision, the consequence is that the system will sends a lot of signals for the same collision. Furthermore I think some kind of filtering on events dispatching is useful, otherwise each listener will receive all the collision events and have to filter them himself.

**Comment by JacopoG** [comment]
I have removed some unnecessary things.

**Comment by DeimosA** [comment]
ye, it depends on how it's handled on the receiving end

**Pull request merged**